lib/decorateWithMetrics.js

```diff
@@ -0,0 +1,132 @@
+import _ from 'underscore';
+
+/**
+ * Each key is a method name and the value is an array of calls metadata
+ * @type {Record<String, Array<CallMeta>>}
+ */
+let stats = {};
+
+/**
+ * Class instances are easily tracked by Dev tools memory snapshots
+ * Wrapping a result with such instance allows us to track it without
+ * incurring extra weight
+ */
+class TaggedItem {
+    constructor(data) {
+        Object.assign(this, data);
+        this.ref = data;
+    }
+}
+
+/**
+ * Capture tags in a weak map
+ * When the original object (key) is no longer used anywhere in app - the data (TaggedItem)
+ * is free for garbage collection and removed
+ * Using dev tools we see which TaggedItems are still retained and what's using them - preventing GC
+ * @type {WeakMap<object, TaggedItem>}
+ */
+const taggedItems = new WeakMap();
+
+/**
+ * Wraps a function with metrics capturing logic
+ * @param {function} func
+ * @param {String} [alias]
+ * @returns {function} The wrapped function
+ */
+function decorateWithMetrics(func, alias = func.name) {
+    if (_.has(stats, alias)) {
+        throw new Error(`"${alias}" is already decorated`);
+    }
+
+    stats[alias] = [];
+
+    function decorated(...args) {
+        const startTime = performance.now();
+        let result;
+
+        const originalReturnValue = func.apply(this, args);
+
+        /*
+        * Promisify the original return value in case the decorated method does not return a promise
+        * Then handlers added here are not affecting the returned promise as they aren't attached to it
+        * */
+        Promise.resolve(originalReturnValue)
+            .then((data) => {
+                result = data;
+
+                // Weak map keys can be objects only
+                if (_.isObject(data)) {
+                    const tag = new TaggedItem(data);
+                    taggedItems.set(data, tag);
+                }
+            })
+            .catch(error => (result = error))
+            .finally(() => {
+                const endTime = performance.now();
+
+                if (!_.has(stats, alias)) {
+                    stats[alias] = [];
+                }
+
+                const methodStats = stats[alias];
+                methodStats.push({
```

Comment:
Looking at some of the output, I only found `methodName, args, duration` valuable. Also... what happened to duration?

Reply:
+1

Except `result` seems valuable since if there are outliers then you might want to consider the data being returned when investigating why something is taking longer than expected.

Reply:
I didn't want to include any computation that can be avoided here
We're just capturing the raw data and then when we actually need to display it it's mapped to take a better shape

On my sample benchmark I use a method like this:

```js
    printMetrics() {
        const data = Onyx.getMetrics().map(call => ({
            method: call.methodName,
            startTime: Math.round(call.startTime),
            endTime: Math.round(call.endTime),
            duration: Math.round(call.endTime - call.startTime),
            args: JSON.stringify(call.args),
            result: JSON.stringify(call.result),
        }));

        // eslint-disable-next-line no-console
        console.table(data);

        Onyx.resetMetrics();
    }
```

https://github.com/kidroca/Expensify.cash/blob/3baa89133035429fdc13424ca2d8a5d67ab435c9/src/libs/Navigation/NavigationRoot.js#L40-L54

---

The times actually have nanosecond precision and they are lengthy float (double) point numbers like `3003.0000141` the value before the dot are milliseconds.